internal/controllers/configuration/konglicense_controller.go

```diff
@@ -0,0 +1,381 @@
+func (r *KongV1Alpha1KongLicenseReconciler) ensureControllerStatusProgrammedCondition(
+	ctx context.Context, license *kongv1alpha1.KongLicense,
+	programmedStatus metav1.ConditionStatus,
+	reason string, message string,
+) error {
+	// Get the latest status of target KongLicense.
+	err := r.Client.Get(ctx, k8stypes.NamespacedName{Name: license.Name}, license)
+	if err != nil {
+		return fmt.Errorf("failed to get latest version of KongLicense %s: %w", license.Name, err)
+	}
+
+	fullControllerName := r.fullControllerName()
+	// Find the managed controller status item and append new item when absent.
+	controllerStatus, controllerIndex, found := lo.FindIndexOf(license.Status.KongLicenseControllerStatuses, func(controllerStatus kongv1alpha1.KongLicenseControllerStatus) bool {
+		return controllerStatus.ControllerName == fullControllerName
+	})
+	if !found {
+		wantedControllerStatus := kongv1alpha1.KongLicenseControllerStatus{
+			ControllerName: fullControllerName,
+			Conditions: []metav1.Condition{
+				{
+					Type:               ConditionTypeProgrammed,
+					LastTransitionTime: metav1.Now(),
+					Status:             programmedStatus,
+					Reason:             reason,
+				},
+			},
+		}
+		license.Status.KongLicenseControllerStatuses = append(license.Status.KongLicenseControllerStatuses, wantedControllerStatus)
+		return r.Client.Status().Update(ctx, license)
+	}
+
+	wantedCondition := metav1.Condition{
+		Type:               ConditionTypeProgrammed,
+		LastTransitionTime: metav1.Now(),
+		Status:             programmedStatus,
+		Reason:             reason,
+		Message:            message,
+	}
+	// Find the "programmed" condition in the condition list.
+	programmedCondition, conditionIndex, found := lo.FindIndexOf(controllerStatus.Conditions, func(condition metav1.Condition) bool {
+		return condition.Type == ConditionTypeProgrammed
+	})
+	if !found {
+		// Return error if the number of conditions already reached the limit.
+		if len(license.Status.KongLicenseControllerStatuses) >= maxConditionNum {
+			return fmt.Errorf("already %d condition items in controller status exceeding the limit %d, cannot add new item",
+				len(license.Status.KongLicenseControllerStatuses), maxConditionNum)
+		}
+		license.Status.KongLicenseControllerStatuses[controllerIndex].Conditions = append(
+			license.Status.KongLicenseControllerStatuses[controllerIndex].Conditions, wantedCondition,
+		)
+		return r.Client.Status().Update(ctx, license)
+	}
+	if programmedCondition.Status != programmedStatus {
+		license.Status.KongLicenseControllerStatuses[controllerIndex].Conditions[conditionIndex] = wantedCondition
+		return r.Client.Status().Update(ctx, license)
+	}
+
+	return nil
+}
+
+// WrapKongLicenseReconcilerToDynamicCRDController wraps KongLicenseReconciler to DynamicCRDController
+// to watch precense of KongLicense CRD to avoid aborts if KongLicense is not installed when controller initialized.
+// REVIEW: Is there a better way to resolve that?
```

Comment:
So far we've been using `DynamicCRDController` only to allow Gateway API CRDs to be installed during runtime (the only non-GW API CRD is `KongUpstreamPolicyReconciler` although it depends on `HTTPRoute` so it's also wrapped in that). 

For regular independent Kong CRDs we were rather expecting the CRD to be installed and cache sync to fail if it's not. As `KongLicense` is introduced as `alpha`-maturity feature, I think it could be irritating to users if they upgrade (forgetting to upgrade CRDs) and get weird crashes because of cache sync timeouts so I'm ok with going with the `DynamicCRDController` for that.

Another option would be to make this feature-gated and enable the controller only if the feature gate is enabled, making it non-breaking for people not using the `KongLicense` straight away.